docs/src/rfcs/002-shader-stage-interpolants.md

```diff
@@ -0,0 +1,188 @@
+| tessellation control output | maxTessellationControlPerVertexOutputComponents / 4 | 128 | 128
+| tessellation evaluation input | maxTessellationEvaluationInputComponents / 4 | 128 | 128
+| tessellation evaluation output | maxTessellationEvaluationOutputComponents / 4 | 128 | 128
+| geometry input | maxGeometryInputComponents / 4 | 64 | 128
+| geometry output | maxGeometryOutputComponents / 4 | 128 | 128
+| fragment input | maxFragmentInputComponents / 4 | 64 | 128
+| fragment output | maxFragmentOutputAttachments | 4 | 8
+
+
+### GLSL Linkage mechanism
+
+Historically GLSL used to link shader stage interpolants by name, however in recent years it acquired the `layout(location = N, component = K)` attribute on interpolants. GLSL also allows partial linking where a previous stage can output more data then the stage before.
+
+For pixel shader outputs in modern GLSL, instead of return a value from a function one writes to a global `out` type. And similarly, shader inputs also come from globals.
+
+### Decorators
+
+Similar to DirectX, Vulkan has the following decorations that can be applied to interpolants.
+
+| Vulkan | DirectX equivalent
+| ---- | ---- |
+| - | linear |
+| Flat| nointerpolation
+| NoPerspective| noperspective
+| Centroid | centroid
+| Sample| sample
+
+Vulkan has some additional explanations about these decorations as well:
+
+> A variable decorated with `Flat` will not be interpolated. Instead, it will have the same value for every fragment within a triangle. This value will come from a single provoking vertex. A variable decorated with `Flat` can also be decorated with Centroid or Sample, which will mean the same thing as decorating it only as `Flat`.
+
+## Metal shading language
+
+__Mostly todo.__
+
+One thing to note as a different to SPIR-V, GLSL and HLSL is that MSL doesn't seem to require explicit binding semantics or `Location` specifiers. It's specification has a `[[attribute(N)]]` specifier, but it's only used on tesselation control points and vertex shader inputs. It's specifically not used for communication between shader stages.
+
+# SPIR-V Practicalities
+
+Vertex shaders allow `Location` and `Component` decorations on input variable declarations. The `Location` decoration specifies which vertex input attribute is used to read and interpret the data that a variable will consume. The `Component` decoration allows the location to be more finely specified for scalars and vectors, down to the individual components within a location that are consumed. The components within a location are 0, 1, 2, and 3. A variable starting at component N will consume components N, N+1, N+2, …​ up through its size. For single precision types, it is invalid if the sequence of components gets larger than 3.
+
+
+## Proposal
+
+```rust
+// these `attribute`'s would match the input assembler setup on the CPU
+struct VertexInput {
+    #[attribute(0)]
+    position: Vec3,
+    #[attribute(1)]
+    normal: Vec3,
+    #[attribute(2)]
```

Comment:
What behavior do you imagine when attributes are missing? Error, implicit default value, something else?

Reply:
I think it's most likely that all of the members of a `#[stage_in]` struct will have to have some kind of decoration - either `#[position]`, `#[vertex_id]`, etc or `#[attribute(..)]`. However, I don't think it's required for `#[stage_in]` to have to have an `#[attribute(..)]` (see below).

Reply:
Potentially #[spirv(attribute="position")] or #[spirv(attribute=0)] instead?

Reply:
The reason suggested much sorter syntax for most of these things is because they are quite common in shader cod and pop up all the time.

@charles-r-earp the word `attribute` here refers to a specific kind of attribute which is modeled in Vulkan by something known as the [vertex input attributes](https://www.khronos.org/registry/vulkan/specs/1.2-extensions/man/html/VkVertexInputAttributeDescription.html), adding position in there would complicate things I feel because that's one of the outputs of a vertex shader.

Reply:
This may be an indicator that the word `attribute` is badly chosen for this case since it's already such an overloaded term, maybe `vertex_attribute` would be better, what do you think?

Reply:
> The reason suggested much sorter syntax for most of these things is because they are quite common in shader cod and pop up all the time.

We've already decided to go with `#[spirv(...)]` in previous discussions, bikeshedding over this isn't useful. If we want to change that decision, let's open a separate MCP with drawbacks/advantages/alternatives and discuss it there.

> However, I don't think it's required for `#[stage_in]` to have to have an `#[attribute(..)]` (see below).

I don't quite understand this comment, could you explain this more? What do you mean by `#[stage_in]` having an `#[attribute(..)]`?